src/MongoDB.Bson/IO/JsonReader.cs

```diff
@@ -1623,6 +1731,77 @@ private BsonValue ParseObjectIdExtendedJson()
             return new BsonObjectId(ObjectId.Parse(valueToken.StringValue));
         }
 
+        private BsonValue ParseRegularExpressionExtendedJsonCanonical()
+        {
+            VerifyToken(":");
+            VerifyToken("{");
+
+            string pattern = null;
+            string options = null;
+
+            var nextToken = PopToken();
+            while (nextToken.Type != JsonTokenType.EndObject)
+            {
+                if (nextToken.Type != JsonTokenType.String && nextToken.Type != JsonTokenType.UnquotedString)
+                {
+                    var message = string.Format("JSON reader expected a string but found '{0}'.", nextToken.Lexeme);
+                    throw new FormatException(message);
+                }
+                var name = nextToken.StringValue;
+
+                VerifyToken(":");
+
+                nextToken = PopToken();
+                if (nextToken.Type != JsonTokenType.String)
+                {
+                    var message = string.Format("JSON reader expected a string but found '{0}'.", nextToken.Lexeme);
+                    throw new FormatException(message);
+                }
+                var value = nextToken.StringValue;
+
+                switch (name)
+                {
+                    case "pattern":
+                        pattern = value;
+                        break;
+                    case "options":
+                        options = value;
```

Comment:
There is also a restriction on the options field: 
`BSON Regular Expression options MUST be in alphabetical order.`
But I don't think that it's required here.